`#![register_{attribute,lint}_tool]`

jyn514 opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rfcs/pull/3808)*

[Rendered](https://github.com/rust-lang/rfcs/blob/master/text/3808-register-tool.md)

Co-authored by @BD103.

Thank you to everyone who gave feedback on early versions of this RFC.

jyn514 committed:
`#![register_{attribute,lint}_tool]`

Co-authored-by: BD103 <59022059+BD103@users.noreply.github.com>

jyn514 committed:
initial comments

- mention that `rustdoc -w json` will include the attributes
- fix semver-checks spelling
- apparently adding `inline` can be a breaking change 🙄

jyn514 commented:
going to go ahead and ping all the authors of extern tools that i know:
- @xd009642 (tarpaulin)
- @kkysen @thedataking (c2rust)
- @Daalbrecht @bd103 (bevy_lint)
- @obi1kenobi (cargo semver-checks)
- @smoelius (dylint)
- @xfrednet (marker)
- @nbdd0121 (klint)
- not sure who maintains kani, maybe @celinval ?
- not sure who maintains prusti, maybe @fpoli ?
- @eddyb (rust-gpu)
- @chorman0773 (lccc)
- @Nadrieril ([charon](https://github.com/AeneasVerif/charon))
- not sure who maintains [crubit](https://github.com/google/crubit), maybe @ssbr ?

jyn514 committed:
oops the second MSRV mention was duplicated and also in the wrong place

obi1kenobi commented:
I would happily use this in `cargo-semver-checks`. Users keep asking for things like "how can I disable lint X on item Y"; right now the only options we can offer them are:
- disable the lint everywhere
- mark the module `#[doc(hidden)]`, thereby making it non-public API

Both of those are very blunt instruments, and thoroughly unsatisfying as an answer to our users. This RFC addresses those problems directly. I'm excited to see it.

jyn514 committed:
mention that #[coverage] could be a tool attribute

epage commented:
> I would happily use this in `cargo-semver-checks`. Users keep asking for things like "how can I disable lint X on item Y"; right now the only options we can offer them are:
> 
>     * disable the lint everywhere
> 
>     * mark the module `#[doc(hidden)]`, thereby making it non-public API
> 
> 
> Both of those are very blunt instruments, and thoroughly unsatisfying as an answer to our users. This RFC addresses those problems directly. I'm excited to see it.

How would you be able to leverage this? irrc rustdoc gives you immediate attributes but not enough other attributes to resolve lint levels.

epage commented on text/3808-register-tool.md:
````diff
@@ -0,0 +1,257 @@
+}
+```
+
+Crate-level lints for external tools can be configured in `Cargo.toml`, like any other crate-level lint:
````
To confirm, this is changing cargo from only allowing specific tools to any?

Do we need to provide a way for people to read this, like `cargo metadata`?

What if tool registration is later added to `Cargo.toml`, should we only limit the `lint` namespace to that? That would likely requre an edition to roll out as it would be a breaking change.

jyn514 replied:
> Do we need to provide a way for people to read this, like cargo metadata?

I think that's up to the cargo team. I am happy to specify it in the RFC but I don't think it's strictly necessary.

> What if tool registration is later added to `Cargo.toml`, should we only limit the `lint` namespace to that? That would likely requre an edition to roll out as it would be a breaking change.

I don't follow. What about that change would be breaking?

BD103 replied:
Right now Cargo prints the following error with this `Cargo.toml`:

```toml
[lints.bevy_lint]
pedantic = "warn"
```

```sh
warning: /Users/bdeep/dev/bevy/bevy_cli/bevy_lint/Cargo.toml: unrecognized lint tool `lints.bevy_lint`, specifying unrecognized tools may break in the future.
supported tools: cargo, clippy, rust, rustdoc
```

The warning hurts the UX of using the `[lints]` table with custom tools, but it doesn't stop the table from being used. The RFC does imply that Cargo should support lint tools without any warnings, which would make adding predefined tools a potentially breaking change. I'm not sure if this is what you're talking about, though.

jyn514 replied:
I misunderstood, you are saying that going from the situation after this RFC (supporting [lints] but not tool registration in Cargo.toml) to the future you are proposing (supporting both, and giving a hard error for lints that didn't register your tool) is a breaking change.

I do not think Cargo should give a hard error if a tool is not registered in Cargo.toml. Partly because that's rustc's job, but partly because there are many different ways to configure crate-level attributes, and we should use the final resolved values of the crate attributes, not just what Cargo knows about. So that avoids this problem altogether.

jyn514 replied:
> imo if the user is using a `lints.<tool>` table without it being registered, it should behave in the exact same way as rustc, whether it errors or something else. Rustc has no visibility into this to do something about it.

i disagree, i think it's exactly the opposite. `lints.tool` tells cargo to pass `-W tool::` flags, at which point rustc has all the info to do something about it. but cargo doesn't have all the info, because it doesn't know the crate level attributes.

> I feel like there is no value in providing support for `lints` in this RFC without a canonical way of accessing the data.

the value is in having a way to configure lints for a workspace without having to duplicate them in the root of each crate.

that said, if you feel strongly about this i will try to come up with a canonical format for `cargo metadata`.

jyn514 replied:
how does this look?
```json
{
  "packages": [/* ... */],
  "lints": {
    "bevy_lint": {
      "lint_name": "allow",
    }
  }
```
for the toml file
```toml
[lints.bevy_lint]
lint_name = "allow"
```

jyn514 replied:
oh, maybe you expect this to show up for each dependency too. we could nest them in that way, it would go under `.packages[*].lints`.

epage replied:
> i disagree, i think it's exactly the opposite. lints.tool tells cargo to pass -W tool:: flags, at which point rustc has all the info to do something about it. but cargo doesn't have all the info, because it doesn't know the crate level attributes.

Why would we convert all tools to rustc flags?  The way the RFC is written, it sounds like the tools would not be running inside of rustc.

> the value is in having a way to configure lints for a workspace without having to duplicate them in the root of each crate.

My concern more is how are tools expected to access the lint levels from `Cargo.toml`?

> how does this look?

lints exist at the
- workspace level (well, they will when we add cargo lints)
- package level

> oh, maybe you expect this to show up for each dependency too. 

We have this weird tension in `cargo metadata` about giving people the information they need without slowing things down.  Lints are going to add a lot to the output.  I wonder if we can skip dependencies but what if a tool does need it?

BD103 replied:
I don't think adding `cargo metadata` support to the lints table should be a concern for this RFC. I don't believe it is useful enough for the amount of bikeshedding it requires.

Tools are free to use `cargo locate-project [--workspace]` to find the crate and workspace manifests, then deserialize the `[lints]` table with `serde` + `toml`. This solution is easy to implement and benefits the projects by letting them read custom lint configuration:

```toml
[lints.my_tool]
my_lint = { level = "warn", foo = 3, bar = false }
```

Not all tools even need this. `rustc` drivers already have the proper flags passed to them as `--warn my_tool::my_lint`, so most of the time they won't even need direct access to `[lints]`.

Adding `cargo metadata` support is a nice-to-have, but by no means should it block this RFC.

jyn514 replied:
i've figured out the confusion here. @epage yes, you're right - passing lint flags to rustc does nothing. we need to pass the lint flags to the *tools* instead. but the external tools aren't controlled by cargo (that's the whole point) and so cargo doesn't know how to pass the flags.

i think the right approach here is
1. strip out all the cargo stuff from this RFC, except maybe tool registration
2. change the "authors of tools" guide to suggest using `[metadata.your_tool.lints]` to configure lints crate-level tools

does that sound reasonable to you?

epage replied:
> change the "authors of tools" guide to suggest using [metadata.your_tool.lints] to configure crate-level lints for tools

I see this as a short term solution.  If we're going with that kind of short-term solution there, should we do the same for registration?

DaAlbrecht replied:
I agree i think it makes more sense to strip out all the cargo stuff for this RFC and do it in a follow up even if registering tools would be a big QoL improvement.

jyn514 replied:
i have moved all the cargo features to "future possibilities". i don't think there is an urgency to land them now.

epage commented on text/3808-register-tool.md:
```diff
@@ -0,0 +1,257 @@
+- Projects may wish to have both a proc-macro crate with lints and a CLI with lints. To allow this, we would require `proc_macro_lint` to create an exhaustive list of lints that can be created, such that we can still run `unknown_lints` and do not need to create a new cooperation mechanism between `proc_macro_lint` and `register_lint_tool`, nor to require users of the project to distinguish the t ...
+- We could allow proc-macros to register a scoped tool, such that e.g. `#[serde::flatten]` is valid while the proc-macro is expanding, but not elsewhere in the crate. This is similar to [derive helpers], but namespaced. We would have to take care to avoid ambiguity between the scoped tool and globally registered tools in such a way that external tools still do not need to perform name resolution.
+- Once [expression attributes] are stabilized, this would also allow tool attributes on expressions.
+- We could support registering tools through `Cargo.toml`, which Cargo would then pass to `rustc` using `--crate-attr` (similar to `--check-cfg`). We would have to consider how this interacts with duplicate registration; we don't want to error if a tool is configured both through Cargo.toml and in source code.
```
Covering every crate in a package can be annoying otherwise which may be worth calling out.

jyn514 replied:
I think I'm going to remove the hard error on duplicate registration and only error if people try to register a built-in tool, which resolves this nicely. I'll also specify a mechanism for Cargo to pass this to rustc. how does `[package] tools = ["c2rust", "bevy_lint"]` sound?

epage replied:
Creating a new top-level table is going to need a lot more elaboration, especially if its as generic as `tools`.

Should this RFC be split into two parts?
- rustc side
- cargo side

epage commented on text/3808-register-tool.md:
```diff
@@ -0,0 +1,257 @@
+
+This runs the risk that external tools will add attributes that change the meaning of the code, such that the behavior is different when the tool is present. There is not much we can do about this other than to ask tool authors not to do that.
+
+This introduces a new "meta-breaking" concern: once this is stabilized, adding a new tool namespace, like we did for `diagnostic` and `rust_analyzer`, becomes a breaking change. I think in practice there will be few enough external tools that this will not be a major concern and we can just pick a name that isn't already in use. Note that, because `![register_*_tool(...)]` for a built-in tool err ...
```
Should we pre-emptively reserve some namespaces?

jyn514 replied:
That seems like a good idea - at least `metadata` is something I'd like to reserve. It's much easier to reserve them now than to try and add them later.

petrochenkov replied:
I'd rather go the opposite way and remove all the hardcoded tools on the next edition, and always require registering them explicitly.
There are zero reasons for rustc to know about, e.g. `rust_analyzer`.

petrochenkov replied:
>at least `metadata` is something I'd like to reserve

This reminded me of something - https://github.com/rust-lang/rust/pull/75802.

jyn514 replied:
oh, "adding a new namespace has to be over an edition boundary" seems very reasonable, yes. that avoids having to treat built-in tools differently from registered tools.

jyn514 commented on text/3808-register-tool.md:
```diff
@@ -0,0 +1,257 @@
+- We could "just not do this". That makes it harder to write external tools, and in practice just means that people use `cfg_attr` instead of a namespace, which seems strictly worse.
+- We could relax the constraint that tool names cannot conflict with local items. This requires tools to do name resolution; but in practice I do not think we can expect tools to do this, and we must assume that the tool will behave differently than the compiler (`rustfmt` already does this today).
+- We could add a syntax to disambiguate tool names from local items. That would add inconsistency with the existing built-in tool attributes, and requires tool authors to parse both the new and existing syntax.
+- We could drop the change to name resolution, such that extern crates continue to take precedence over tools. This allows no way to use a tool and a crate with the same name, unless we add a new syntax.
```
to be clear, you are suggesting that the following be a hard error?
```rust
#![register_attribute_tool(pyo3)]
extern crate pyo3;
#[pyo3::pymodule]  // error: ambiguous between tool and crate
struct Foo {}
```
I think that would not be ideal, but I guess I could live with it as long as `#[::pyo3::pymodule]` is still allowed. I feel somewhat strongly that `#[::pyo3::pymodule]` should not give an error, the `extern crate as` workaround is not very well known. even if people use it as `my_pyo3 = { package = "pyo3" }` in Cargo.toml it's still a bit indirect.

cc @mejrs

petrochenkov replied:
>to be clear, you are suggesting that the following be a hard error?

Yes, just because if it's an error initially, then in the future we'll be able to backward compatibly do any of the resolution changes from this RFC (precedence, subnamespaces) if really necessary.

`#[::pyo3::pymodule]` should certainly work, there's no ambiguity there, `::foo` only looks into extern prelude.

jyn514 replied:
I thought of another complication. Say we want to add more built-in tools in the future, or to reserve a tool namespace today. If adding a built-in tool causes breakage because it makes crate names ambiguous, suddenly it's much harder to extend the namespace.

I guess we could have different rules for built-in tools than for registered tools, but that kinda sucks.

mejrs replied:
> I think that would not be ideal, but I guess I could live with it as long as `#[::pyo3::pymodule]` is still allowed.

I've thought about this and and having ` #[::pyo3::pymodule]` work in this case but not `#[pyo3::pymodule]` seems like a bad idea to me. I think it'd be better if they *both* errored or if both worked. If we allow the former then it's gonna be one of those gotchas that *everyone* runs into. "Oh yeah you have to prefix the macro path with `::` because otherwise it's ambiguous between the tool and the macro". I think I'd like to have them named the same but it might be better if authors are forced to choose different names.

mejrs replied:
> you as the author can choose a different name for your tool

I do not disagree in principle, but I'm just a bit paranoid of the situation where someone chooses the crate name as their tool name without thinking about it very much and it ends up being awkward later on. Imagine a situation like:
- author creates a library (without proc macros)
- author creates a tool attribute with the same name, this works fine because of no proc macros
- author later decides to add proc macros
- (without prefix `::`) these macros collide with the tool name
- the author would have to decide to either rename the tool or tell the users to prefix the macros. The latter sounds like a noob trap.

jyn514 replied:
I don't think we can prevent this situation. The last thing still happens even if we prevent disambiguating the tool, we just *force* you to rename, we are strictly limiting your options.

Consider a different case: someone writes a crate, someone else writes a tool, they don't know about each other. you combine them and they conflict. I don't think we should force you in this case to use `extern crate overlap as my_overlap`, that seems strictly worse than allowing you to disambiguate at the call site.

tmandry replied:
I think the RFC was updated to "always error" so @petrochenkov's original concern is resolved for now. Is that right?

jyn514 commented on text/3808-register-tool.md:
```diff
@@ -0,0 +1,257 @@
+
+The `#![register_attribute_tool(ident)]` crate-level attribute adds a new tool to the tool attribute prelude. The `#![register_lint_tool(ident)]` crate-level attribute adds a new tool to the lint prelude. For both attributes, tools must be a single ident, not a nested path.
+
+`#[no_implicit_prelude]` keeps its current behavior, i.e. it removes all tools that are in the prelude by default. However, those tools can now be explicitly added back with e.g. `#![register_attribute_tool(rustfmt)]`.
```
I will specify that `no_implicit_prelude` does not affect tools.

jyn514 commented:
> How would you be able to leverage this? irrc rustdoc gives you immediate attributes but not enough other attributes to resolve lint levels.

this is not a rustdoc problem. this is a language level problem. semver-checks would have the same problem if it were using `syn` to parse; today, there is simply no place to put the attributes where rustc doesn't hard error.

jyn514 commented:
oh, i misunderstood, you were asking how semver-checks would use this *if* it were implemented. i would expect semver-checks to reimplement the precedence rules for lints. it doesn't have CLI flags, which means `[lints]` in Cargo.toml won't work, that's unfortunate. maybe rustdoc-json could expose those flags.

jyn514 committed:
Specify cargo registration

- make cargo registration the main user interface
- specify that `no_implicit_prelude` does not affect tools

jyn514 committed:
fix typo

jyn514 commented on text/3808-register-tool.md:
````diff
@@ -0,0 +1,257 @@
+```
+This is in order to not require external tools to perform name resolution. This restriction may be relaxed in the future to favor tool names.
+
+The `#![register_attribute_tool(ident)]` crate-level attribute adds a new tool to the tool attribute prelude. The `#![register_lint_tool(ident)]` crate-level attribute adds a new tool to the lint prelude. For both attributes, tools must be a single ident, not a nested path.
````
I don't see why not. `--crate-attr` is effectively a built-in macro, and I explicitly want to allow that. Using other macros seems fine as long as they expand to a crate-level attribute.

petrochenkov replied:
I think that should already be covered by usual restricted shadowing rules.

jyn514 replied:
> #![foo::bar]

to be clear: are you asking about a proc-macro defined in a crate called `foo`? that's disallowed today, proc-macros can't be crate-level attributes.

if you mean a tool named `foo` then that's an inert attribute, not a macro.

petrochenkov replied:
(In any case crate-level macros are still extremely unstable.)

jyn514 committed:
update summary

jyn514 committed:
rework name resolution semantics

- all ambiguity is forbidden, not just tool<->local ambiguity. as a
  result, there is no change to precedence order.
- no special-casing for attribute macros; tools are visible in any
  context in the type namespace
- the suggested way to add a new built-in tool is "wait for an edition"

celinval commented on text/3808-register-tool.md:
```diff
@@ -0,0 +1,287 @@
+
+Like today, attributes and lints in a tool namespace are always considered used by the compiler. The compiler does not verify the contents of any tool attribute, except to verify that all attributes are syntactically valid [tool attributes].
+
+Registering a predefined tool (`clippy`, `miri`, etc.) using `#![register_*_tool(...)]` is an error.
```
Where will this list be maintained?

Also, today you can't add any attribute that starts with `rustc`. Will that still be the case?

celinval replied:
From what I understand, today you cannot add a tool attribute that includes the prefix `rustc`, for example, you can register a tool `rustc_foo`, but the compiler will reject the attribute `#[rustc_foo::bar]`.

jyn514 replied:
oh, i see. this is actually not coming from the register_tool code, this is the same code that rejects `#[rustc_baz]`.

i think it is fairly reasonable to keep this reserved, but i can make sure it's rejected earlier in `register_tool` so that people aren't confused.

petrochenkov replied:
>but i can make sure it's rejected earlier in register_tool so that people aren't confused.

I don't think it makes sense to prevent this, there are many other ways to make an unusable attribute that are not rejected, like `mod rustc_something { pub use my_attribute; }` or `use my_attribute as rustc_something;`, but we do not prohibit modules or imports containing `rustc` preventively.

celinval replied:
As long as we are consistent. Either it works without any hacks or it fails early on.

celinval reviewed (commented):
It's awesome to see this moving forward. Thanks!

For Kani, it's better to include @zhassan-aws / @carolynzech.

jyn514 committed:
specify that the rustc tool namespace is reserved

carolynzech commented on text/3808-register-tool.md:
````diff
@@ -0,0 +1,289 @@
+```
+For each `NAME`:
+- If `lints` is `true`, Cargo passes [`--crate-attr`]`=register_lint_tool(NAME)`
+- If `attributes` is `true`, Cargo passes `--crate-attr=register_lint_tool(NAME)`
````
```suggestion
- If `attributes` is `true`, Cargo passes `--crate-attr=register_attribute_tool(NAME)`
```

carolynzech reviewed (commented):
Thank you!

celinval commented on text/3808-register-tool.md:
```diff
@@ -0,0 +1,289 @@
+[summary]: #summary
+
+This RFC adds two new attributes:
+- `#![register_lint_tool(tool_name)]` allows controlling namespaced lints with `#[warn(tool_name::lint_name)]`
+- `#![register_attribute_tool(tool_name)]` allows using tool names in [inert attributes][inert] with `#[tool_name::attribute_name(token_tree)]`.
```
May I suggest we use `#![register_tool_{attribute|lint}` instead? I was also wondering if we could have a `#[register_tool]` which would add the tool to both namespaces, the attribute and lint.

BD103 replied:
I worry that `register_tool_{attribute|lint}` will imply we are registering _individual_ attributes or _individual_ lints, when that is not the case. If someone sees the attribute `#![register_tool_lint(...)]`, I worry they'll do this:

```rust
#![register_tool_lint(bevy::panicking_methods)]
#![register_tool_lint(bevy::missing_reflect)]
#![register_tool_lint(bevy::duplicate_bevy_dependencies)]
// ...
```

I think the name `register_lint_tool` much better signals that it's for registering a _tool_, not a _lint_ or _attribute_.

celinval replied:
Gotcha... That makes sense. Thanks

kennytm replied:
unless we allow nested tool name like `com::example::tool`[^1], I don't think `#![register_tool_lint(bevy::panicking_methods)]` will compile because the tool name can only be an `$:ident` not a `$:path`, so it is actually not an issue.

[^1]: `#![warn(com::example::tool::lint)]` is interpreted as having a tool name `com`.

jyn514 replied:
sure, but why not avoid the confusion in the first place?

kennytm replied:
> sure, but why not avoid the confusion in the first place?

IMO `#![register_tool(clippy)]` is also confusing, it sounds like it will make `rustc` run the "tool" `clippy` by "registering" it to the compilation process. The standard reference refers to this as "tool *name*" rather than just "tool". I think `#![register_tool_name(clippy)]` would be clear, but `#![register_attribute_tool_name(clippy)]` is quite verbose.

jyn514 replied:
That's not what "register" means. I think there is *some* amount of confusion possible, but I don't think it's likely enough that we need to make all the names longer. If people are adding these attributes, it's because the tool's docs told them to, and the docs should also say how to run the tool. If people are reading these attributes in existing code, then, well, I don't think it's that much more confusing than any other attribute in the language? `#[inline]` especially does not mean what most people think it means.

ssbr commented:
> not sure who maintains crubit, maybe @ssbr ?

(Yup). FWIW, the main downside with this proposal is that it requires some extra work on the user's POV.

Here's the current Crubit workflow:

* You depend on a [crubit_annotate crate](https://github.com/google/crubit/blob/main/support/crubit_annotate.rs), which has a bunch of attribute-macros which will check that you have well-formed inputs, and expand to a structured doc-comment.
* Crubit itself parses the doc-comment.

(We used to use tool attributes, before @cramertj created this new setup.)

Now, with the proposal, AIUI, to have a similar workflow using tool attributes, the user must both depend on the attribute macro crate, as well as independently register the tool attribute which the proc macro will generate. This is unfortunate -- the user needs to do two things, and needs to know the name of the tool attribute even if they never spell it.

I like the proc macro wrapper, because it can force compile-time errors even if you happen to not be running the tool at the moment. The metadata Crubit consumes is not necessarily simple and so does need some validation. My preference as far as user experiences go are that a proc macro can generate inert attributes without them being registered by the crate that uses the proc macro, or else that there's a globally-available tool attribute, similar to [`clang::annotate`](https://clang.llvm.org/docs/AttributeReference.html#annotate), so that we could generate something like `rust::tool_attribute(crubit::attribute_name, extra, args, go, here)`. That way, the only thing the user needs to do is depend on the attribute macro crate, and not worry about what it does past that.

I'm hoping this helps. FWIW, we have a working process, and I think this proposal is certainly compatible with many paths forward that lead to what I want, and definitely doc-comments are a bit... egh. Thanks for all your work here, and thanks for soliciting more feedback! I really appreciate it.

jyn514 commented on text/3808-register-tool.md:
```diff
@@ -0,0 +1,289 @@
+
+- Proc macros wish to register custom lints; see [`proc_macro_lint`]. We would have to establish some mechanism to prevent overlapping namespaces. Perhaps `warn(::project::lint_name)` could refer to the proc macro and `warn(project::lint_name)` would refer to any registered tool (only when a `project` tool is regisetered; in the common case where no tool is registered, `project::` would still ref ...
+- Projects may wish to have both a proc-macro crate with lints and a CLI with lints. To allow this, we would require `proc_macro_lint` to create an exhaustive list of lints that can be created, such that we can still run `unknown_lints` and do not need to create a new cooperation mechanism between `proc_macro_lint` and `register_lint_tool`, nor to require users of the project to distinguish the t ...
+- We could allow proc-macros to register a scoped tool, such that e.g. `#[serde::flatten]` is valid while the proc-macro is expanding, but not elsewhere in the crate. This is similar to [derive helpers], but namespaced. We would have to take care to avoid ambiguity between the scoped tool and globally registered tools in such a way that external tools still do not need to perform name resolution.
```
@ssbr i'm going to respond to you here so we can have things in threads; i think derive helpers are basically what you're asking for.

> Now, with the proposal, AIUI, to have a similar workflow using tool attributes, the user must both depend on the attribute macro crate, as well as independently register the tool attribute which the proc macro will generate. This is unfortunate -- the user needs to do two things, and needs to know the name of the tool attribute even if they never spell it.

so, this reminds me strongly of [derive macro helpers](https://doc.rust-lang.org/nightly/reference/procedural-macros.html#derive-macro-helper-attributes). i double checked just now and derive helpers *are* [preserved by macro-expansion](https://github.com/user-attachments/assets/136f9052-5017-4495-b7b5-e61c2e420f8b). so i expect you to continue to be able to use your proc-macro after this is stabilized, it would just use a different (non-namespaced) syntax. that seems ok - extending derive helpers to a namespace is a nice addition, but i don't think it's strictly necessary for you to keep using this. i don't see a need for proc macros to use the same syntax as hand-written code.

ssbr replied:
Are derive macro helpers also able to be emitted by, well, non-derive macros? They should work, I don't think we tried them.

>  i don't see a need for proc macros to use the same syntax as hand-written code.

+1

jyn514 replied:
i'm not sure if they are today, but that seems like a very reasonable extension RFC. i am not going to write that RFC but you are welcome to and i can give you pointers on how to get started.

i'll add this to future extensions.

jyn514 committed:
add `register_tool` as an alias

jyn514 committed:
move all the cargo stuff to future possibilities

jyn514 committed:
mention attribute helpers as an extension

epage commented on text/3808-register-tool.md:
```diff
@@ -0,0 +1,271 @@
+
+The Rust language can be extended and analyzed using external tools. If your tool can parse Rust, you may wish to allow configuring it at sub-crate levels (e.g. individual functions, types, and modules). To reuse the same syntax as the official tools, like Clippy and Rustfmt, instruct your users to add `#![register_lint_tool(your_tool)]` (if your tool only adds new lints) or `#![register_attribut ...
+
+We do not specify a syntax for crate-level configuration. We suggest using `[package.metadata.your_tool]` in Cargo.toml.
```
```suggestion
We do not specify a syntax for package-level configuration. We suggest using `[package.metadata.your_tool]` in Cargo.toml for now.  See [Future Possibilities]
```
(technically the attribute is crate-level and yes our terminology is confusing)

epage commented on text/3808-register-tool.md:
```diff
@@ -0,0 +1,271 @@
+[future-possibilities]: #future-possibilities
+
+- We could allow registering tools in Cargo.toml (with a `package.tools` field?). This would avoid duplicating tool registration for each crate in the workspace. This depends on [`--crate-attr`] being stabilized.
+- We could make `[lints]` support external tools as a first-class feature. This needs a way for tools to read the metadata out of `Cargo.toml` (e.g. `cargo metadata`, or just parsing the toml file), because cargo does not drive external tools. Additionally, we cannot guarantee that tools will actually read the metadata.
```
We could have a `[lints]` entry automatically register the tool, even if the table is empty.

epage commented on text/3808-register-tool.md:
```diff
@@ -0,0 +1,271 @@
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+- We could allow registering tools in Cargo.toml (with a `package.tools` field?). This would avoid duplicating tool registration for each crate in the workspace. This depends on [`--crate-attr`] being stabilized.
```
We should also consider this process allowing registering tool-specific `cfg`s (e.g. `cfg(kani)`)

We might want to consider a way for tools to define these entries and for users to pull them in somehow.

A tools table may have enough overlap with [runnable dependencies](https://github.com/rust-lang/cargo/issues/2267) that we may want to develop one with the other in mind.

We need to consider whether this table is a `workspace` table like `[profiles]` or a package table.  For the former, we should auto-copy it to the package on publish like we do `workspace.resolver`.  If the latter, we should consider whether workspace inheritance should be allowed.

jyn514 replied:
> We should also consider this process allowing registering tool-specific cfgs (e.g. cfg(kani))

This seems reasonable, but not all tools support cfgs, so doing it for all tools defeats the point of check-cfg. So we would need some way to configure the behavior.

I would be extremely surprised to see someone building an external tool as a dependency. I can mention it I guess but I don't know that it makes sense to go out of our way to support it.

I don't have strong opinions on the other things, but I will mention them in future possibilities.

kennytm commented on text/3808-register-tool.md:
```diff
@@ -0,0 +1,271 @@
+
+Unknown tool names in lints remain a hard error until the story for proc-macro lints is resolved (see [Future possibilities](#future-possibilities)).
+
+`#![no_implicit_prelude]` does not affect tools.
```
What does this sentence mean? 

(1) If by "tool" you mean "tool name", today `#![no_implicit_prelude]` does skip the tool prelude and thus remove all the tool names from the attributes (*but not lints*, perhaps a bug), so the "tool names" do get affected.
```rust
#![no_implicit_prelude]
#[deny(clippy::identity_op)] // unexpected, won't raise E0710 (unknown tool name)
#[rustfmt::skip] // expected, error[E0433]: failed to resolve: use of unresolved module or unlinked crate `rustfmt`
const X: u8 = 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 +0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0;
```

(2) If by "tool" you mean "tool invocation" then yes, the tool does not need to respect `#![no_implicit_prelude]`. For instance `rustfmt` recognizes the `#[rustfmt::skip]` here just fine. But tools that rely on `rustc` such as `clippy` will still be affected by E0433.

(3) Or "tool" here just mean the `register_tool` call and that the following behaves the same with or without `#![no_implicit_prelude]`?

```rust
#![no_implicit_prelude]
#![register_tool(bevy)] // always ok
#![register_tool(rustc)] // always error
```

jyn514 replied:
I do mean 1), tool name. The RFC is specified in terms of the behavior that will be stabilized, not in terms of the current implementation.

kennytm commented on text/3808-register-tool.md:
```diff
@@ -0,0 +1,271 @@
+Registering a predefined tool (`clippy`, `miri`, etc.) using `#![register_*_tool(...)]` is an error.
+
+The `rustc` tool namespace is currently reserved and will continue to be reserved after this RFC, i.e, `register_*_tool(rustc)` is an error.
+Additionally, any tool namespace starting with `rustc`, such as `rustc_lint`, is reserved.
```
does this include names like `rustcat`

jyn514 replied:
yeah. i am not a fan of the current attribute gating but i don't want to touch it in this rfc.

danielhenrymantilla commented:
Very nice! 🙌

---

I'm not understanding 100% the reasoning behind denying the ambiguous inert tool attribute _vs._ a genuine attribute, although maybe my confusion stems from `rustfmt::skip` being rather special / _purely_ syntactical (although I could envision there being other purely-syntactical markers by some "cheap" tools or whatnot, right?).

Consider the following:

```rs
mod rustfmt {
    pub use ::tokio::main as skip;
}

#[rustfmt::skip]
async fn main()
             {}
```

  - this, currently, does compile perfectly fine;
  - and `rustfmt` does see the `rustfmt::skip` and does not reformat that `fn main`.

A tool (inert) attribute is just that, an inert / nothing-doing syntactical marker that the compiler is to let slide. But a genuine attribute, is just as much a syntactical marker as an inert attribute; so I do wonder about non-compiler-name-res-using tools such as `rustfmt` not making a difference between the two.

But I imagine that if a tool were to hook itself onto compiler machinery so as to actually involve name res and _semantics_, then a difference between a genuine attribute and an inert one could be made.

In which case, denying the ambiguity makes sense.

But all this may hint at some mention of the distinction between semantics-capable tools (_e.g._, able to resolve `cfg`s), _vs._ purely-syntactical ones being warranted? 🙂

jyn514 commented on text/3808-register-tool.md:
````diff
@@ -0,0 +1,271 @@
+The `rustc` tool namespace is currently reserved and will continue to be reserved after this RFC, i.e, `register_*_tool(rustc)` is an error.
+Additionally, any tool namespace starting with `rustc`, such as `rustc_lint`, is reserved.
+
+Ambiguity between a tool name and any other name in the type namespace is always a hard error. For example, this code would error:
+
+```rust
+#![register_tool(name)]
+
+extern crate name;
+#[name::skip] // ERROR: ambiguous
+#[::name::skip] // OK
+fn foo() {}
````
> I'm not understanding 100% the reasoning behind denying the ambiguous inert tool attribute vs. a genuine attribute
> A tool (inert) attribute is just that, an inert / nothing-doing syntactical marker that the compiler is to let slide. But a genuine attribute, is just as much a syntactical marker as an inert attribute; so I do wonder about non-compiler-name-res-using tools such as rustfmt not making a difference between the two.

> But I imagine that if a tool were to hook itself onto compiler machinery so as to actually involve name res and semantics, then a difference between a genuine attribute and an inert one could be made.

> In which case, denying the ambiguity makes sense.

@danielhenrymantilla i think you are imagining that the *tool* is responsible for checking there is no ambiguity. that is not the case. the *compiler* is responsible for checking there is no ambiguity, precisely because we want to allow tools that don't do name-res to still use attributes. it's ok if tools accept more code than the compiler, but i think it would be quite bad if the tool interpreted the code differently than the compiler (this is the same reason i have a big "please don't use attributes that change the meaning of the code" mention in the "tool author guide" section).

> But all this may hint at some mention of the distinction between semantics-capable tools (e.g., able to resolve cfgs), vs. purely-syntactical ones being warranted? 🙂

i think distinguishing here is unnecessary complexity, and if we can get away with it we should just use the same rules for all of them. the exception here is between attribute_tool and lint_tool: since they don't affect name res for each other, we can make it easier to add lint-only tools that have the same name as an existing crate without causing breakage. but i don't think we should add more types of tools than necessary.

jyn514 committed:
extend cargo future possibilities

jyn514 committed:
don't directly prohibit tools starting with rustc

BD103 commented:
> Just asking. Is it possible to implement any custom "attribute -> extra-codegen" using this feature? Such `register_tool(llvm)` for example to add llvm-attrs to codegen/IR. I really need some _probably-always-unstable_ llvm attributes like `noprofile`, `skipprofile` and `no_sanitize(“address”)`/`disable_sanitizer_instrumentation`.

From my understanding, no. Registered tool attributes do not affect compilation using `rustc`, they're only there to be parsed by 3rd-party tools. If you want to introduce an attribute that affects how the code is compiled, you'll need to introduce that into the Rust compiler itself.

tmandry commented on text/3808-register-tool.md:
```diff
@@ -0,0 +1,279 @@
+- We could allow proc-macros to register a scoped tool, such that e.g. `#[serde::flatten]` is valid while the proc-macro is expanding, but not elsewhere in the crate. This is similar to [derive helpers], but namespaced. We would have to take care to avoid ambiguity between the scoped tool and globally registered tools in such a way that external tools still do not need to perform name resolution.
+- Once [expression attributes] are stabilized, this would also allow tool attributes on expressions.
+- Some existing attributes, such as [`coverage`], have exactly the semantics of a tool attribute: they add additional meaning when a specific feature or flag is enabled, and ignored otherwise. They could use this mechanism (over an edition boundary, as described above).
+
```
```suggestion
- We can allow defining tool attributes in crates as inert declarative macros, which would enable IDE tooling to provide suggestions and syntax-level checking of the attributes. Taking advantage of this would require users to move away from register_tool and use a crate dependency with the name of the tool instead.
  - One danger of accepting this proposal now is that tools will use the crate name for their binary instead of saving it for an eventual support library that can declare attribute macros and potentially other things. This could be mitigated with explicit guidance to use another crate name. For example, kani uses [kani-verifier](https://crates.io/crates/kani-verifier) for the binary while reserving [kani](https://crates.io/crates/kani) for an eventual library.
  - This is somewhat akin to the way the [proc_macro_lint](https://github.com/rust-lang/rust/pull/135432) proposal puts declared lints in the macro namespace.
- We can allow tool attributes to be declared using Rust types. This would the use of versioned, structured metadata that can be shared among different tools and reflection APIs.
  - This can be combined with tool attribute macros to present more interesting syntax options.

```

tmandry commented:
@rfcbot fcp merge lang

This RFC would allow code to use tool-defined attributes and lints without relying on nightly features. These are used during development with custom tools that are specially configured to run on that codebase. The RFC proposes the simplest thing that can work for this purpose, and indeed one that exists unstably in the compiler today.

### Future possibilities

I have suggested some [future possibilities](https://github.com/rust-lang/rfcs/pull/3808#discussion_r2600619092) that make sense to me, noting some drawbacks that came up in the [Zulip discussion](https://rust-lang.zulipchat.com/#narrow/channel/213817-t-lang/topic/namespaced.20tool.20attrs/with/562544172). These would have to overcome implementation complexity involving name resolution in attributes, or find a way to avoid doing that.

While I find these possibilities appealing, I don't think we should let the perfect be the enemy of the good. First, because we might not need it. Second, because there aren't significant ecosystem effects of someone adopting `#![register_tool]` in their own codebase. If we release a more fancy version of this feature in the future, users will be free to migrate to it independent of their dependencies or dependents. In the less-common case where a tool operates on attribute metadata from multiple crates in a crate graph, it should still be possible to define a migration story that supports both schemes simultaneously.

jyn514 committed:
Propose a future extension to allow rustc to check the syntax of a tool attribute

Co-authored-by: Tyler Mandry <tmandry@gmail.com>

jyn514 commented on text/3808-register-tool.md:
```diff
@@ -0,0 +1,286 @@
+# Summary
+[summary]: #summary
+
+This RFC adds two new attributes:
```
fixed, thanks

jyn514 committed:
fix typos, minor clarifications

jyn514 committed:
rewrite the "guide" section to reflect that ambiguity is always an error

jyn514 committed:
mention that custom inner attributes are still banned

tomassedovic commented:
@tmvkrpxl0 I'm not sure what's wrong with the website, but the document can be viewed here: https://github.com/rust-lang/rust-project-goals/blob/main/src/2024h2/Contracts-and-invariants.md

jyn514 commented on text/3808-register-tool.md:
```diff
@@ -0,0 +1,310 @@
+but we need to know the current scope in order to give an ambiguity error.
+
+`#![no_implicit_prelude]` does not affect tools,
+i.e. the tool prelude is always present.
```
That’s correct, yes. The motivation is that tools (e.g. rustfmt) already ignore this distinction in practice, so this brings the language into consistency with the implementation.

(Fixing this in rustfmt would be … challenging, since it doesn’t do name resolution.)

jyn514 committed:
clarify that not respective `no_implicit_prelude` is a breaking change

jyn514 committed:
fix broken link

jyn514 committed:
fix typo

ehuss committed:
Fix register_tool headers

This updates section headers to use Markdown 2nd level headings. As part
of https://github.com/rust-lang/rfcs/pull/3883 we switched the template
to not use level-1 headings
